File: AlgorithmicToolbox/week3_greedy_algorithms/1_money_change/1_money_change.cpp
#include <iostream>

/*
This program implements an elementary greedy algorithm used by cashiers all over the
world millions of times per day.
The goal is to find the minimum number of coins needed to change the input value 
(an integer) into coins with denominations 1, 5, and 10.
*/

int get_change(int m) {
	//write your code here
	int arr[3]{ 0 };
	int den1{ 10 }, den2{ 5 }, den3{ 1 }, sum{ 0 };

	if (m == 0) return 0;

	if ((m / den1) > 0) {
		arr[0] = m / den1;
		m = m % den1;
	}
	if ((m / den2) > 0) {
		arr[1] = m / den2;
		m = m % den2;
	}
	if ((m / den3) > 0) {
		arr[2] = m / den3;
	}
	for (int i = 0; i < 3; ++i) {
		sum = sum + arr[i];
	}

	return sum;
}

int main() {
	int m;
	std::cin >> m;
	std::cout << get_change(m) << std::endl;
}

File: AlgorithmicToolbox/week3_greedy_algorithms/1_money_change/1_money_change.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


# Money change
# python3

"""
This program implements an elementary greedy algorithm used by cashiers all over the world
millions of times per day.
The goal is to find the minimum number of coins needed to change the input value (an integer)
into coins with denominations 1, 5, and 10.

"""

def money_change(money):
    assert 0 <= money <= 10 ** 3
    
    coins_list = [0]*3
    
    if money == 0:
        return 0
    
    den1, den2, den3 = 10, 5, 1
    if money//den1 > 0:
        coins_list[0] = money//den1
        money = money%den1
    if money//den2 > 0:
        coins_list[1] = money//den2
        money = money%den2
    if money//den3 > 0:
        coins_list[2] = money
    return sum(coins_list)
    


if __name__ == '__main__':
    input_money = int(input())
    print(money_change(input_money))


# In[ ]:




